### DISCUSSION TITLE
Async business rule causing SaveObjectAsync to return false

### DISCUSSION THREAD

**Discussion by CROprogrammer** [MarimerLLC/csla#3336] [Questions] [answered]
In my BusinessBase object I have added async business rule which checks that my object is unique. Because of that, I have noticed that in my MVC application first time when I add new object (in my case role) everything works fine. Then other times when I try to create/edit object, SaveObjectAsync returns false because my object is marked busy because of my business rule. Any tip how to avoid that problem so that my object is not busy when SaveObjectAsync is being called?

**Comment by rockfordlhotka** [comment]
There's a `CheckRulesAsync` method that is specifically designed to make it easy to check all rules and wait until they complete (including async rules) - specifically to support the MVC and similar server-side models.

There is an `OnValidationComplete` event that is raised when all rules are complete.

I also seem to remember that we implemented something that you could await when `IsBusy` becomes `false` - specifically to address this issue. Or maybe it was just a code pattern to follow - hmm, it has been a while.

**Reply by rockfordlhotka** [reply]
The more I think about it, the more I'm convinced that the pattern I mentioned about awaiting `IsBusy` being `false` is that `CheckRulesAsync` method.

**Reply by CROprogrammer** [reply]
Where should `CheckRulesAsync` be called? 
In my MVC controller I have the next code:
```
public async Task<IActionResult> Create()
{
  RoleList roles = await _roleListDataPortal.FetchAsync();
  RoleEdit roleEdit = roles.AddNew();

  return View(roleEdit);
}

[HttpPost]
[ValidateAntiForgeryToken]
public async Task<IActionResult> Create(RoleEdit roleEdit)
{
  try
  {
    if (await SaveObjectAsync(roleEdit, false))
      return RedirectToAction(nameof(Index));
    else
      return View(roleEdit);
  }
  catch
  {
    return View(roleEdit);
  }
```
First I create empty Role object, and then I edit it in its create page. After I submit its form then I'm redirected to second method `public async Task<IActionResult> Create(RoleEdit roleEdit)` in which `SaveObjectAsync` is called, and inside that method object's `SaveAsync` is called which checks if object `IsBusy`.
Should I override default `SaveAsync` method and call `CheckRulesAsync` in it? Or should I call it somewhere else?

**Reply by rockfordlhotka** [reply, marked as answer]
What is happening here, is that the MVC data binding creates the `RoleEdit` instance and loads the properties of the object. That runs the rules as the properties are set. 

When I say "MVC data binding", in reality this is handled by a CSLA-specific implementation in [CslaModelBinder](https://github.com/MarimerLLC/csla/blob/main/Source/Csla.Web.Mvc.Shared/CslaModelBinder.cs).

Looking at the implementation, I can see where it might be wise to rework this to call `CheckRulesAsync` instead of `CheckRules` and entirely abstract the issue into the binder. This change would probably address your issue without any special code in any individual controller.

My recommendation to you is to take a copy of `CslaModelBinder`, switch the `CheckRules` call, and configure aspnetcore to use this binder instead of the original.

**Reply by rockfordlhotka** [reply]
Recorded as an issue in #3338 - probably for CSLA 8.

**Reply by richardbartley** [reply]
I think you can also get this when you have a Parent -> Child/children scenario and the child has async business rules and you (me) forgets to override the IsValid property on the parent to also consider the child/children.

**Comment by CROprogrammer** [comment]
I ended up overriding `SaveAsync` method in RoleEdit class like this:

```
protected override async Task<RoleEdit> SaveAsync(bool forceUpdate, object userState, bool isSync)
    {
      await BusinessRules.CheckRulesAsync();

      return await base.SaveAsync(forceUpdate, userState, isSync);
    }
```

**Reply by rockfordlhotka** [reply]
And that solved the issue?

If so, that implies that the change I'm proposing for CSLA 8 should be a good solution.

**Reply by CROprogrammer** [reply]
Yup, it now works as I wanted it to
